rfcs/0049-flakes.md

```diff
@@ -0,0 +1,706 @@
+  `github:edolstra/nixpkgs/release-19.03`. You can use such locations
+  ("flake references") directly in the `nix` command:
+
+  ```
+  # nix build github:NixOS/patchelf
+  ```
+
+* For a reproducible result, you can also use a specific revision:
+
+  ```
+  # nix build nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06:hello
+  ```
+
+* To get information about a flake:
+
+  ```
+  # nix flake info nixpkgs
+  Description:   A collection of packages for the Nix package manager
+  Revision:      a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ...
+  ```
+
+* In addition to the global registry, there is a per-user
+  registry. This can be used to pin flakes to the current version:
+  ```
+  # nix flake pin nixpkgs
+  ```
+  or to a specific version:
+  ```
+  # nix flake add nixpkgs github:edolstra/nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ```
+
+* Flakes can have dependencies on other flakes. For example, [the
+  `patchelf`
+  flake](https://github.com/NixOS/patchelf/blob/master/flake.nix)
+  depends on the `nixpkgs` flake. To ensure reproducibility,
+  dependencies are pinned to specific versions using a *lock file*
+  (e.g. [for
+  patchelf](https://github.com/NixOS/patchelf/blob/master/flake.lock)). Lock
+  files are generated automatically.
+
+* Flakes are evaluated in pure mode, meaning they can't access
+  anything other than their own source or declared dependencies. This
+  allows `nix` to cache evaluation results.
+
+* The `nix` command is flake-based. For example,
+  ```
+  # nix build
+  ```
+  builds the flake in the current directory.
+
+# Motivation
+[motivation]: #motivation
+
+Flakes are motivated by a number of serious shortcomings in Nix:
+
+* While Nix pioneered reproducible builds, sadly, Nix expressions are
+  not nearly as reproducible as Nix builds. Nix expressions can access
+  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
+  environment variables, and Git repositories. This means for instance
+  that it is hard to ensure reproducible evaluation of NixOS or NixOps
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
+
+* There is no standard way to compose Nix projects. Typical ways are
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems. It can be
+seen as a "Cargo/NPM/... for Nix" except that it's built into Nix
+rather than a separate tool.
```

Comment:
Let me insist on something because I read some really odd things while going back in the discussion:

>thinking about making nix a package manager

Nix is *already* a package manager. Just because it doesn't conform with what you are used to with *some* package managers like npm and Cargo does not make it less a package manager. Package managers have an almost 30-year long history, with very varied design decisions. And even today, npm and Cargo (and RubyGems and PyPI) do not represent the only way to make a good modern package manager.

Reply:
Isn't the point of this comparison that flakes would make it easier to package language-specific package manager ecosystems for nix-the-non-language-specific-package-manager?

As an example, currently there's a couple GHCs in nixpkgs, if you need one that isn't you're out of luck. Flakes would decouple the GHCs needed by nixpkgs from the ones that can be made available. Flakes'd be more like stackage LTSes, available for as long as anyone needs them. (Note that I'm not saying it's impossible to pin nixpkgs to the latest revision that had a GHC or that other technical solutions exist. I am saying flakes'd turn this from a PITA into simply specifying the right flake.)

Reply:
Sure, it would make such a use simpler and this is one of the great things about flakes.

But this sentence as is can be interpreted in many ways. If everyone who reads it interprets it differently, it's better to just remove it. This RFC is overall very clear, but this sentence does not help.

Reply:
Perhaps it'd be better to rephrase as "Cargo/NPM/... for the [Nix Expression Language](https://nixos.wiki/wiki/Nix_Expression_Language)"?

Otherwise as @Zimmi48 mentions this is conflated with Nix The Package Manager(this is something that also got me confused).

Also, I wonder if there is some shorthand the community uses to refer to the Nix Expression Language(NixEx?).

Reply:
@steve-chavez, nix-lang is the name of the irc channel so I'd go with that.

As your understanding of the paragraph is almost completely orthogonal to mine (In my understanding flakes only happen to be able to provide a way to distribute nix libraries kind of by accident, their main merit is to modularize package repositories.) I now strongly agree it needs rephrasing.

Reply:
> Perhaps it'd be better to rephrase as "Cargo/NPM/... for the [Nix Expression Language](https://nixos.wiki/wiki/Nix_Expression_Language)"?
> 
> Otherwise as @Zimmi48 mentions this is conflated with [Nix The Package Manager](https://nixos.wiki/wiki/Nix_Package_Manager)(this is something that also got me confused).

If we can’t even describe what it is we want to create except by using the names of other things, that should give us pause. We can’t say on the website “Flakes are npm for Nix” and expect to be taken seriously.

Reply:
That doesn't mean similes can't be useful.

Reply:
That’s true, but we still haven’t come up with a better generic term for it than “package manager”.

Reply:
@toonn I definitely meant "package manager for nix-lang" in my initial proposal that got turned into flakes.

@alyssais I don't think the fact of naming difficulty here implies what you're thinking. The fact is, we have a programming language whose primary domain is managing packages, it's unsurprising that we have to be a bit verbose/analogous when trying to refer to the role of the language applied to the language itself.

Reply:
The RFC could be more clear that this is about `nix-lang`. Some parts also mention `Nix expressions`, which is not elucidating. Perhaps also rename those to `nix-lang expressions`? .

Reply:
Imo, Nix expression is probably the clearest way of describing this. It unambiguously refers to an expression written in the Nix Expression Language.

Reply:
I do think that if we had more names / "brands", it would make the layering concerns I commented on easier to resolve.

Reply:
I've removed this sentence.